src/query/service/src/interpreters/interpreter_table_drop.rs

```diff
@@ -54,92 +57,93 @@ impl Interpreter for DropTableInterpreter {
+            Err(error) => {
+                if (error.code() == UNKNOWN_TABLE
+                    || error.code() == UNKNOWN_CATALOG
+                    || error.code() == UNKNOWN_DATABASE)
+                    && self.plan.if_exists
+                {
+                    return Ok(PipelineBuildResult::create());
+                } else {
+                    return Err(error);
+                }
+            }
+        };
 
-        if tbl.is_none() && !self.plan.if_exists {
-            return Err(ErrorCode::UnknownTable(format!(
-                "Unknown table `{}`.`{}` in catalog '{}'",
-                db_name, tbl_name, catalog_name
+        let engine = tbl.get_table_info().engine();
+        if matches!(engine, VIEW_ENGINE | STREAM_ENGINE) {
+            return Err(ErrorCode::TableEngineNotSupported(format!(
+                "{}.{} engine is {} that doesn't support drop, use `DROP {} {}.{}` instead",
+                &self.plan.database,
+                &self.plan.table,
+                engine,
+                engine,
+                &self.plan.database,
+                &self.plan.table
             )));
         }
-        if let Some(tbl) = tbl {
-            let engine = tbl.get_table_info().engine();
-            if matches!(engine, VIEW_ENGINE | STREAM_ENGINE) {
-                return Err(ErrorCode::TableEngineNotSupported(format!(
-                    "{}.{} engine is {} that doesn't support drop, use `DROP {} {}.{}` instead",
-                    &self.plan.database,
-                    &self.plan.table,
-                    engine,
-                    engine,
-                    &self.plan.database,
-                    &self.plan.table
-                )));
-            }
-            let catalog = self.ctx.get_catalog(catalog_name).await?;
+        let catalog = self.ctx.get_catalog(catalog_name).await?;
 
-            // drop the ownership
-            let tenant = self.ctx.get_tenant();
-            let role_api = UserApiProvider::instance().get_role_api_client(&self.plan.tenant)?;
-            let db = catalog.get_database(&tenant, &self.plan.database).await?;
-            role_api
-                .drop_ownership(&GrantObjectByID::Table {
-                    catalog_name: self.plan.catalog.clone(),
-                    db_id: db.get_db_info().ident.db_id,
-                    table_id: tbl.get_table_info().ident.table_id,
-                })
-                .await?;
+        // drop the ownership
+        let tenant = self.ctx.get_tenant();
+        let role_api = UserApiProvider::instance().get_role_api_client(&self.plan.tenant)?;
+        let db = catalog.get_database(&tenant, &self.plan.database).await?;
+        role_api
+            .drop_ownership(&GrantObjectByID::Table {
+                catalog_name: self.plan.catalog.clone(),
+                db_id: db.get_db_info().ident.db_id,
+                table_id: tbl.get_table_info().ident.table_id,
+            })
+            .await?;
```

Comment:
Perhaps we should move this section to come after `catalog.drop_table_by_id`? This would prevent the situation that if something goes wrong in between, the ownership is dropped, but the table still exists.

Reply:
Yes, I think so, let me add comments and change this.